drivers/sensor/akm09918c/akm09918c.c

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright (c) 2023 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT asahi_kasei_akm09918c
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/logging/log.h>
+
+#include "akm09918c.h"
+#include "akm09918c_reg.h"
+
+LOG_MODULE_REGISTER(AKM09918C, CONFIG_SENSOR_LOG_LEVEL);
+
+static int akm09918c_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	struct akm09918c_data *data = dev->data;
+	const struct akm09918c_config *cfg = dev->config;
+	uint8_t buf[9] = {0};
+
+	if (chan != SENSOR_CHAN_ALL && chan != SENSOR_CHAN_MAGN_X && chan != SENSOR_CHAN_MAGN_Y &&
+	    chan != SENSOR_CHAN_MAGN_Z && chan != SENSOR_CHAN_MAGN_XYZ) {
+		LOG_WRN("Invalid channel %d", chan);
+		return -EINVAL;
+	}
+
+	if (data->mode == AKM09918C_CNTL2_PWR_DOWN) {
+		if (i2c_reg_write_byte_dt(&cfg->i2c, AKM09918C_REG_CNTL2,
+					  AKM09918C_CNTL2_SINGLE_MEASURE) != 0) {
+			LOG_ERR("Failed to start measurement.");
+			return -EIO;
+		}
+
+		/* Wait for sample */
+		LOG_DBG("Waiting for sample...");
+		k_busy_wait(AKM09918C_MEASURE_TIME_US);
```

Comment:
this is a relatively long busy wait

Reply:
I also thought it was odd but used the other `ak` sensor as reference. I'm happy to change it but assumed there was a hard reason for it.

Reply:
Yeah, let's convert that to a sleep

Reply:
Done